pf/tfbridge/provider.go

```diff
@@ -95,45 +98,70 @@ func ShimProviderWithContext(ctx context.Context, p pfprovider.Provider) shim.Pr
 func newProviderWithContext(ctx context.Context, info tfbridge.ProviderInfo,
 	meta ProviderMetadata) (pl.ProviderWithContext, error) {
 	const infoPErrMSg string = "info.P must be constructed with ShimProvider or ShimProviderWithContext"
-	if info.P == nil {
-		return nil, fmt.Errorf("%s: cannot be nil", infoPErrMSg)
-	}
-	schemaOnlyProvider, ok := info.P.(*schemashim.SchemaOnlyProvider)
-	if !ok {
-		return nil, fmt.Errorf("%s: found non-conforming type %T", infoPErrMSg, info.P)
-	}
-	p := schemaOnlyProvider.PfProvider()
 
-	server6, err := newProviderServer6(ctx, p)
-	if err != nil {
-		return nil, fmt.Errorf("Fatal failure starting a provider server: %w", err)
-	}
-	resources, err := pfutils.GatherResources(ctx, p)
-	if err != nil {
-		return nil, fmt.Errorf("Fatal failure gathering resource metadata: %w", err)
-	}
+	var (
+		server6            tfprotov6.ProviderServer
+		resources          pfutils.Resources
+		datasources        pfutils.DataSources
+		providerConfigType *tftypes.Object
+	)
+	switch p := info.P.(type) {
+	case *schemashim.SchemaOnlyProvider:
```

Comment:
There is some code smell here with the case analysis, perhaps the capability this code is seeking is missing an interface.

Reply:
We could convert to an interface.

```go
type PFProviderServer interface {
	Server()      tfprotov6.ProviderServer
	Resources()   pfutils.Resources
	DataSources() pfutils.DataSources
	Config()      *tftypes.Object
}
```

I think that would help clean up some existing code that does the same thing. Good idea.

Reply:
Bonus points if pfutils.* is not there or is simplified. It might be profitable to revisit and retroactively review the design of that.